Update Pod overview (#25756)
kubernetes/website: content/en/docs/concepts/workloads/pods/_index.md

system semantics, and makes it feasible to extend the cluster's behavior without
changing existing code.

## Pod update and replacement

As mentioned in the previous section, when the Pod template for a workload

> resource is changed, the controller will create new Pods based on the updated

Reviewer: ```suggestion
resource is changed, the controller creates new Pods based on the updated
```

template instead of updating or patching the existing Pods.

Kubernetes doesn't prevent you from managing Pods directly. It is possible to

> update a Pod in place. However, the Pod update operation has some limitations:

Reviewer: ```suggestion
update some fields of an existing Pod, in place. However, the Pod update operation has some limitations:
```
?

> update a Pod in place. However, the Pod update operation has some limitations:

Reviewer 2: nit: Does it make sense to replace "in place" with: 
It is possible to update a running Pod.

nit: What exactly is the Pod "update operation"?
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#patch-pod-v1-core

Author: update include 'patch' and 'replace', revised.

- Most of the metadata about a Pod is immutable. For example, you cannot

>   change the `namespace`, `name`, `uid`, `creationTimestamp` fields, you can

Reviewer: ```suggestion
  change the `namespace`, `name`, `uid`, `creationTimestamp` fields; you can
```

>   change the `namespace`, `name`, `uid`, `creationTimestamp` fields, you can

Reviewer 2: nit?: For example, you cannot change the `namespace`, `name`, `uid`, or `creationTimestamp` fields.
nit: The `generation` field is unique. This field only accepts updates that increment the field's current value?

  only increment the `generation` field if you want to change it.
- If the `metadata.deletionTimestamp` is set, no new entry can be added to the

>   `metadata.finalizers` list.

Reviewer: Maybe mention that this is not specific to Pods (but! I can't think of some concise wording).

Author: Maybe this is already done in the GC concept/task.

- Pod updates may not change fields other than `spec.containers[*].image`,
  `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or

>   `spec.tolerations`. Only adding new item into `spec.tolerations` is
>   permitted.

Reviewer: ```suggestion
   `spec.tolerations`. For `spec.tolerations`, you can only add new entries.
```

- When updating the `spec.activeDeadlineSeconds` field, two types of updates

>   are allowed:

Reviewer 2: @sftim suggestions look good. Does the `nil` setting imply that the field is unset or possibly set to 0, ""?
You can change the `spec.activeDeadlineSeconds` field by:
- modifying the field from `nil` to a positive number, or
- modifying the field from a positive number ...

Author: nil is unset.

Reviewer: nil is a bit surprising. On the wire it's JSON and the field would be _I think_ `null` or just not present. If you used `kubectl edit replicaset/foo` (say) then you'd get YAML and again there's either `null` or you omit the field.

In the source code for the API server, it's `nil`. But: Kubernetes tries to be implementation neutral; if someone wants to write a 100% conformant Kubernetes API server in Python, more strength to them.

>   1. update the field from nil to a positive value;
>   1. update the field from a positive value to a smaller, non-negative value.

Reviewer: ```suggestion
  1. setting the field (updating the field from nil to a positive number);
  1. updating the field from a positive number to a smaller, non-negative number.
```
?

system semantics, and makes it feasible to extend the cluster's behavior without
changing existing code.

## Pod update and replacement

As mentioned in the previous section, when the Pod template for a workload
resource is changed, the controller creates new Pods based on the updated
template instead of updating or patching the existing Pods.

Kubernetes doesn't prevent you from managing Pods directly. It is possible to

> update some fields of an running Pod, in place. However, the Pod update

Reviewer 2: nit: a running Pod ...

like 
[`patch`](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#patch-pod-v1-core), and
[`replace`](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#replace-pod-v1-core)
have some limitations:

- Most of the metadata about a Pod is immutable. For example, you cannot
  change the `namespace`, `name`, `uid`, or `creationTimestamp` fields;
  the `generation` field is unique. It only accepts updates that increment the
  field's current value.
- If the `metadata.deletionTimestamp` is set, no new entry can be added to the
  `metadata.finalizers` list.
- Pod updates may not change fields other than `spec.containers[*].image`,
  `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or
  `spec.tolerations`. For `spec.tolerations`, you can only add new entries.
- When updating the `spec.activeDeadlineSeconds` field, two types of updates
  are allowed:

>   1. setting the unssigned field to a positive number; 

Reviewer 2: nit: s/unssigned/unassigned/

>   1. setting the field (updating the field from nil to a positive number);

Reviewer 2: nit: Yes, I think this case could be described as:
`setting the field (or updating the unassigned field) to a positive number;`

>   1. updating the field from a positive nubmer to a smaller, non-negative

Reviewer 2: nit: s/nubmer/number/
